### PR TITLE
Assignment7+8 linq

### PR DESCRIPTION
**Pull request by Bhopss152**
Joshua Hess
j0shh3ss

### CODE REVIEW

**Comment by arose-13** [comment]
(In Luigi accent) Ima gonna review this.

File: Assignment/Assignment.Tests/SampleDataTests.cs [hunk 1]
```diff
@@ -0,0 +1,28 @@
...
+{
+    [TestClass]
+    public class SampleDataTests
+    {
+        [TestMethod]
+        public void TestGetUniqueSortedListOfStatesGivenCsvRows()
+        {
+            // Arrange
+            //Must change the filePath to match your path to the people.csv
+            string filePath = "C:\\Users\\bhopk\\OneDrive\\Documents\\.next\\EWU-CSCD371-2023-Winter\\Assignment\\Assignment\\People.csv";
```
Comment (arose-13) [inline Assignment/Assignment.Tests/SampleDataTests.cs:18]: You need to include this file in you test project by adding it to the project: Right-click on test project -> "Add" -> "Existing Item..." -> select `People.csv`

Don't forget to set it to "Copy if Newer".

File: Assignment/Assignment.Tests/NodeTests.cs [hunk 2]
```diff
@@ -0,0 +1,111 @@
...
+            var result = node.GetEnumerator();
+            result.MoveNext();
+            int firstValue = result.Current;
+            result.MoveNext();
+            int secondValue = result.Current;
+            result.MoveNext();
+            int thirdValue = result.Current;
+
+            // Assert
+            Assert.AreEqual(1, firstValue);
```
Comment (arose-13) [inline Assignment/Assignment.Tests/NodeTests.cs:29]: Don't forget to use the generic version, like this:
```suggestion
            Assert.AreEqual<int>(1, firstValue);
```

**Review by arose-13 — COMMENTED** [review]
1. Implement the ISampleData.CsvRows property, loading the data from the People.csv file and returning each line as a single string. ✔
* Change the "Copy to" property on People.csv to "Copy if newer" so that the file is deployed along with your test project. ❌
* Using LINQ, skip the first row in the People.csv. ✔
* Be sure to appropriately handle resource (IDisposable) items correctly if applicable (and it may not be depending on how you implement it). ✔

2. Implement IEnumerable<string> GetUniqueSortedListOfStatesGivenCsvRows() to return a sorted, unique list of states. ✔
* Use ISampleData.CsvRows for your data source. ✔
* Don't forget the list should be unique. ✔
* Sort the list alphabetically. ✔
* Include a test that leverages a hardcoded list of addresses. ❌
* Include a test that uses LINQ to verify the data is sorted correctly (do not use a hardcoded list). ❌

3. Implement ISampleData.GetAggregateSortedListOfStatesUsingCsvRows() to return a string that contains a unique, comma separated list of states. ✔
* Use ISampleData.GetUniqueSortedListOfStatesGivenCsvRows() for your data source. ✔
* Consider "selecting" only the states and calling ToArray() to retrieve an array of all the state names. ✔
* Given the array, consider using string.Join to combine the list into a single string. ✔

4. Implement the ISampleData.People property to return all the items in People.csv as Person objects ✔
* Use ISampleData.CsvRows as the source of the data. ✔
* Sort the list by State, City, and Zip. (Sort the addresses first then select). ✔
* Be sure that Person.Address is also populated. ✔
* Adding null validation to all the Person and Address properties is optional.
* Consider using ISampleData.CsvRows in your test to verify your results. ❌

5. Implement ISampleDate.FilterByEmailAddress(Predicate<string> filter) to return a list of names where the email address matches the filter. ✔
* Use ISampleData.People for your data source. ✔

6. Implement ISampleData.GetAggregateListOfStatesGivenPeopleCollection(IEnumerable<IPerson> people) to return a string that contains a unique, comma-separated list of states. ✔
* Use the people parameter from ISampleData.People property for your data source. ✔
* At a minimum, use the System.Linq.Enumerable.Aggregate` LINQ method to create your result. ✔
* Don't forget the list should be unique. ✔
* It is recommended that, at a minimum, you use ISampleData.GetUniqueSortedListOfStatesGivenCsvRows to validate your result.

7. Given the implementation of Node in Assignment5
* Implement IEnumerable<T> to return all the items in the "circle" of items. ✔
** Add an IEnumberable<T> ChildItems(int maximum) method to Node that returns the remaining items with a maximum number of items returned less than maximum. ✔

## Fundamentals
Place all shared project properties into a Directory.Build.Props file.
Place all shared project items into a Directory.Build.targets file.
Ensure nullable reference types is enabled ✔
Ensure that you turn on code analysis for all projects(EnableNETAnalyzers) ✔
Set LangVersion and the TargetFramework to the latest released versions available (preview versions optional) ✔
and enabled .NET analyzers for both projects ✔
For this assignment, consider using Assert.AreEqual<T>() (the generic version) ❌
All of the above should be unit tested ❌
Choose simplicity over complexity ✔

## Summary
Overall you did pretty descent, but unfortunately I cannot verify a lot of the functionality that this code is supposed to have due to the lack of tests. `Node` is very well tested, but `SampleData` is not and the only test for it actually fails for me. You also need to add `People.csv` to your test project by right clicking on the test project, hovering over "Add", and select "Existing Item...", and then you can include that file. Be sure to set it to "Copy if Newer" in the advanced options. This should also help your test for SampleData pass.

Other than that, I think you did a pretty good job 😄

File: Assignment/Assignment.Tests/SampleDataTests.cs [hunk 3]
```diff
@@ -0,0 +1,28 @@
...
+            SampleData sampleData = new SampleData(filePath);
+
+            // Act
+            var result = sampleData.GetUniqueSortedListOfStatesGivenCsvRows();
+
+            // Assert
+            Assert.IsNotNull(result);
+        }
+    }
+}
```
Comment (mmoran314) [inline Assignment/Assignment.Tests/SampleDataTests.cs:28]: Missing tests for requirements 2-6.

File: Assignment/Assignment/SampleData.cs [hunk 4]
```diff
@@ -1,30 +1,74 @@
...
+            {
+                IEnumerable<Person> people = from person in CsvRows
+                                             .OrderBy(state => state.Split(',')[6])
+                                             .ThenBy(city => city.Split(',')[5])
+                                             .ThenBy(zip => zip.Split(',')[7])
+                                             .ToList()
+
+                                             let personAttribute = person.Split(',')
+                                             select new Person(personAttribute[1], personAttribute[2],
+                                             new Address(personAttribute[4], personAttribute[5], personAttribute[6], personAttribute[7]),
```
Comment (mmoran314) [inline Assignment/Assignment/SampleData.cs:58]: Consider assigning personAttributes to their own variable for improved readability

**Review by mmoran314 — COMMENTED** [review]
Implement the ISampleData.CsvRows property, loading the data from the People.csv file and returning each line as a single string. ✔

Change the "Copy to" property on People.csv to "Copy if newer" so that the file is deployed along with your test project. ✔
Using LINQ, skip the first row in the People.csv. ✔
Be sure to appropriately handle resource (IDisposable) items correctly if applicable (and it may not be depending on how you implement it). ✔
Implement IEnumerable<string> GetUniqueSortedListOfStatesGivenCsvRows() to return a sorted, unique list of states. ✔

Use ISampleData.CsvRows for your data source. ✔
Don't forget the list should be unique. ✔
Sort the list alphabetically. ✔
Include a test that leverages a hardcoded list of addresses. ❌
Include a test that uses LINQ to verify the data is sorted correctly (do not use a hardcoded list). ❌
Implement ISampleData.GetAggregateSortedListOfStatesUsingCsvRows() to return a string that contains a unique, comma separated list of states. ✔

Use ISampleData.GetUniqueSortedListOfStatesGivenCsvRows() for your data source. ✔
Consider "selecting" only the states and calling ToArray() to retrieve an array of all the state names. ✔
Given the array, consider using string.Join to combine the list into a single string. ✔
Implement the ISampleData.People property to return all the items in People.csv as Person objects ✔

Use ISampleData.CsvRows as the source of the data. ✔
Sort the list by State, City, and Zip. (Sort the addresses first then select). ✔
Be sure that Person.Address is also populated. ✔
Adding null validation to all the Person and Address properties is optional.
Consider using ISampleData.CsvRows in your test to verify your results. ❌
Implement ISampleDate.FilterByEmailAddress(Predicate<string> filter) to return a list of names where the email address matches the filter. ✔

Use ISampleData.People for your data source. ✔
Implement ISampleData.GetAggregateListOfStatesGivenPeopleCollection(IEnumerable<IPerson> people) to return a string that contains a unique, comma-separated list of states. ✔

Use the people parameter from ISampleData.People property for your data source. ✔
At a minimum, use the System.Linq.Enumerable.Aggregate` LINQ method to create your result. ✔
Don't forget the list should be unique. ✔
It is recommended that, at a minimum, you use ISampleData.GetUniqueSortedListOfStatesGivenCsvRows to validate your result.
Given the implementation of Node in Assignment5

Implement IEnumerable<T> to return all the items in the "circle" of items. ✔
Add an IEnumberable<T> ChildItems(int maximum) method to Node that returns the remaining items with a maximum number of items returned less than maximum.

Place all shared project properties into a Directory.Build.Props file.
Place all shared project items into a Directory.Build.targets file.
Ensure nullable reference types is enabled ✔
Ensure that you turn on code analysis for all projects(EnableNETAnalyzers) ✔
Set LangVersion and the TargetFramework to the latest released versions available (preview versions optional) ✔
and enabled .NET analyzers for both projects ✔
For this assignment, consider using Assert.AreEqual<T>() (the generic version) ❌
All of the above should be unit tested ❌
Choose simplicity over complexity ✔

File: Assignment/Assignment.Tests/SampleDataTests.cs (hunk 1 above)
[the remark this answers is not in the document]
Comment (mmwoodfo) [reply Assignment/Assignment.Tests/SampleDataTests.cs:18]: Yes, good suggestion. I don't have this file path on my machine. So these tests fail.

File: Assignment/Assignment/SampleData.cs [hunk 5]
```diff
@@ -1,30 +1,74 @@
...
+            set
+            {
+                _CsvRows = value.Where(x => x != null).Skip(1).ToList();
+            }
+        }
+
+        public IEnumerable<string> GetUniqueSortedListOfStatesGivenCsvRows()
+        {
+            List<string> split = CsvRows.Select(x => x.Split(',')[6]).Distinct().ToList();
+            split.Sort();
```
Comment (mmwoodfo) [inline Assignment/Assignment/SampleData.cs:38]: You should use OrderBy here

File: Assignment/Assignment.Tests/SampleDataTests.cs [hunk 6]
```diff
@@ -0,0 +1,85 @@
...
+using System.Text;
+using System.Threading.Tasks;
+using Microsoft.VisualStudio.TestTools.UnitTesting;
+
+namespace Assignment.Tests
+{
+    [TestClass]
+    public class SampleDataTests
+    {
+        [TestMethod]
```
Comment (mmwoodfo) [inline Assignment/Assignment.Tests/SampleDataTests.cs:13]: You should have a test that uses a hard coded list in here

**Pull request closed without merging**